Fix get_xerr for Data1DInt when all data is filtered

DougBurke opened the pull request:
# Summary

The get_xerr method of Data1DInt would fail if all bins had been ignored and it now returns the empty list.

# Details

I wanted to work on #1129 so added some tests and hit a bug. So this fixes the bug as well as adds a few tests which will be useful for #1129. Not that get_xerr is indirectly called by any routine that plots the data (shown by the to_plot method) so this also fixes that problem.

This PR is the first in a sequence: #1208, #1215, #1216, #1219 (hopefully this is all).

# Example

Before this PR we have

```python
>>> from sherpa.data import Data1DInt
>>> d = Data1DInt('x', [1,2,3], [2,3,4], [1,1,1])
/home/dburke/anaconda/envs/ciao414b1/lib/python3.8/site-packages/sherpa/data.py:50: UserWarning: Converting array [1, 1, 1] to numpy array.
  warnings.warn("Converting array {} to numpy array.".format(array))
>>> d.get_xerr(True)
array([1, 1, 1])
>>> d.ignore(0, 4)
>>> d.get_xerr(True)
ValueError: not enough values to unpack (expected 2, got 1)
>>>  d.to_plot()
IndexError: tuple index out of range
```

The last get_xerr call is the error, and it leads to the to_plot error. The behavior is now

```python
>>> from sherpa.data import Data1DInt
>>> d = Data1DInt('x', [1,2,3], [2,3,4], [1,1,1])
/home/dburke/sherpa/sherpa-master/sherpa/data.py:88: UserWarning: Converting array [1, 1, 1] to numpy array.
  warnings.warn("Converting array {} to numpy array.".format(array))
>>> d.get_xerr(True)
array([1, 1, 1])
>>> d.ignore(0, 4)
>>> d.get_xerr(True)
array([], dtype=float64)
>>> >>> d.to_plot()
(array([], dtype=float64), array([], dtype=int64), None, array([], dtype=float64), 'x', 'y')
```

hamogu reviewed (changes requested):
I've put in a few inline comments, but I do have a more important issue with this solution: It is asymmetric compared to other, similar data classes:
```
xs = numpy.asarray([2, 5, 10, 12, 15, 17, 20])
ys = numpy.ones(xs.size)
d = Data1D('data', xs, ys)
d.ignore(None, None)
print(d.get_xerr(True))
```
gives `None`. Now, `None` is not the same as `[]`. I admit that is was not symmetric before, but if we are *fixing* this, we should make sure we fix it to be consistent. I'm not sure which solution is the *correct* solution, but I strongly suspect that the correct solution is one that gives the same answer for all data classes.

Another small comment: Special casing "len(indep)=1" at this point feels a little hacky. That does not mean it's wrong, it just means that I don't know if it is. My feeling is the `get_evaluation_indep` (or the functions deeper down the stack) should deal with this an return a consistent answer all the way up the chain. Unfortunately, I've never seen (or forgotten) the design documents for this evaluation space/data class hierarchy. So, I don't want to declare this the wrong approach - but I can't say it's the right approach either. I'll try to dig deeper in the code, but I know that this is a very deep 🐰 🕳️ .

Possibly related: #1174

DougBurke commented on sherpa/astro/tests/test_astro_data2.py:
```diff
@@ -85,6 +85,68 @@ def test_need_numpy_channels():
+                         [("channel", '1:10', []),
+                          ("channel", '', [(False, 1, 10)]),
+                          ("channel", '2:9', [(True, 2, 9)]),
+                          ("channel", '2:3,7:9', [(True, 2, 9), (False, 4, 6)]),
```
All these tests are doing is to test the current behavior (since I don't think we have a principled design document to explain what they do now). You'd have to jump ahead to #1219 to check what it is changed to ....

hamogu replied:
Understood. I was not asking for a change, just for clarification in which order things are happening.

DougBurke commented on sherpa/astro/tests/test_astro_data2.py:
```diff
@@ -85,6 +85,68 @@ def test_need_numpy_channels():
+    # Use an ARF to create a channel to energy mapping
+    # The 0.2-2.2 keV range maps to 5.636-61.992 Angsrom
+    #
+    egrid = 0.2 * np.arange(1, 12)
```
But I actually like the use of arange here as it is more closely tied to channels rather than just being a way to create a set of numbers !

hamogu replied:
Either is fine!

DougBurke commented on sherpa/astro/tests/test_astro_data2.py:
```diff
@@ -85,6 +85,68 @@ def test_need_numpy_channels():
+                          ("wave", '6.5:8.3,13.8:24.8', [(True, 6.5, 25), (False, 9.1, 12)]),
+                          ("wave", '5.9:9.5,13.8:24.8', [(True, 6.5, 25), (False, 9.1, 12), (True, 1, 10)]),
+                          ("wave", '6.5:8.3,11.3:41.3', [(True, 6.5, 25), (False, 9.1, 12), (True, 12, 70)]),
+                          ("wave", '5.9:41.3', [(True, 6.5, 25), (False, 9.1, 12), (True, 1, 70)]),
```
There are some tests added in later PRs (not in this particular test though I think).

hamogu replied:
You could just repeat two or three of these here with a `None` instead of the max/min values, e.g. 
`("wave", '', [(False, None, 70)])`, `("wave", '', [(False, 0, None)])`

DougBurke commented:
Random notes on `get_xerr`: I think it's badly named - as far as I can see it's not an x "error", since we don't deal with errors on the independent axis. I think it's more like a "width" .

For `Data1D`, `get_xerr` always seem to return `None` (which would make sense since we don't have "widths" here) which I suggest means we can do different things for `Data1DInt` and `DataPHA`, such as returning `[]`:

```python
>>> from sherpa.data import Data1D
>>> d = Data1D('x', [1, 3, 9], [1, 1, 1])
/home/dburke/anaconda/envs/ciao414b1/lib/python3.8/site-packages/sherpa/data.py:50: UserWarning: Converting array [1, 1, 1] to numpy array.
  warnings.warn("Converting array {} to numpy array.".format(array))
>>> d.get_xerr() is None
True
>>> d.get_xerr(True) is None
True
>>> d.get_xerr(False) is None
True
```

This is not to say that there aren't some inconsistencies here (I note different behaviors for `get_filter` in the `DataPHA` case when no data is selected for example) but I'm not sure this is the PR to be fixing them. For future reference (and this is before #1208 .. #1219 have been applied):

```python
>>> from sherpa.astro.data import DataPHA
>>> import numpy as np
>>> pha = DataPHA('', np.arange(1, 5), np.ones(4))
>>> pha.get_x()
array([1, 2, 3, 4])
>>> pha.get_xerr()
array([1., 1., 1., 1.])
>>> pha.get_y()
array([1., 1., 1., 1.])
>>> pha.get_filter()
'1:4'
>>> pha.ignore()
>>> pha.get_filter()
'No noticed bins'
>>> pha.get_x()
array([1, 2, 3, 4])
>>> pha.get_xerr()
array([1., 1., 1., 1.])
>>> pha.get_y()
array([1., 1., 1., 1.])
>>> pha.get_x(True)
array([1, 2, 3, 4])
>>> pha.get_xerr(True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dburke/anaconda/envs/ciao414b1/lib/python3.8/site-packages/sherpa/astro/data.py", line 3220, in get_xerr
    elo = self.apply_filter(elo, self._min)
  File "/home/dburke/anaconda/envs/ciao414b1/lib/python3.8/site-packages/sherpa/astro/data.py", line 2315, in apply_filter
    return super().apply_filter(self.apply_grouping(data, groupfunc))
  File "/home/dburke/anaconda/envs/ciao414b1/lib/python3.8/site-packages/sherpa/data.py", line 827, in apply_filter
    return self._data_space.filter.apply(data)
  File "/home/dburke/anaconda/envs/ciao414b1/lib/python3.8/site-packages/sherpa/data.py", line 441, in apply
    raise DataErr('notmask')
sherpa.utils.err.DataErr: mask excludes all data
>>> pha.get_y(True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dburke/anaconda/envs/ciao414b1/lib/python3.8/site-packages/sherpa/astro/data.py", line 3194, in get_y
    val = self.apply_filter(val)
  File "/home/dburke/anaconda/envs/ciao414b1/lib/python3.8/site-packages/sherpa/astro/data.py", line 2315, in apply_filter
    return super().apply_filter(self.apply_grouping(data, groupfunc))
  File "/home/dburke/anaconda/envs/ciao414b1/lib/python3.8/site-packages/sherpa/data.py", line 827, in apply_filter
    return self._data_space.filter.apply(data)
  File "/home/dburke/anaconda/envs/ciao414b1/lib/python3.8/site-packages/sherpa/data.py", line 441, in apply
    raise DataErr('notmask')
sherpa.utils.err.DataErr: mask excludes all data
```

DougBurke commented:
For the `len(indep)=1` hack, I agree it's not great but I didn't have the energy to dig down to see what it "should" be returning. The one thing I will say is that it works with our test data...

DougBurke commented on sherpa/astro/tests/test_astro_data2.py:
```diff
@@ -85,6 +85,68 @@ def test_need_numpy_channels():
+    chtype is channel, energy, or wavelength
+    expected is the expected response
+    args is a list of 3-tuples of (flag, loval, hival) where
+    flag is True for notice and False for notice; they define
```
I will probably add a commit to fix this and the get the 0-handling in some of the tests.

DougBurke commented:
I believe I've addressed the comments.

hamogu commented:
Report from SDS meeting: The SDS/Sherpa team has discussed this and agreed that these changes should be done. This does not replace an explicit code review of *how* the changes are done, but we agree that the UI as written here, incl. changes to the current behavior, are matching SDS expectations.

hamogu reviewed (approved):
All concerns from the code review have been addressed or the current approach has been sufficiently explained. I recommend merging this now.

wmclaugh commented on sherpa/astro/tests/test_astro_data2.py:
```diff
@@ -85,6 +85,95 @@ def test_need_numpy_channels():
+    pha = DataPHA('data', chans, counts)
+
+    # Use an ARF to create a channel to energy mapping
+    # The 0.2-2.2 keV range maps to 5.636-61.992 Angsrom
```
```suggestion
    # The 0.2-2.2 keV range maps to 5.636-61.992 Angstrom
```

wmclaugh commented on sherpa/tests/test_data.py:
```diff
@@ -1247,3 +1247,120 @@ def test_data1d_notice_errors_out_on_string_range(lo, hi, emsg, ignore):
+
+    expected is the expected response
+    args is a list of 3-tuples of (flag, loval, hival) where
+    flag is True for notice and False for notice; they define
```
```suggestion
    flag is True for notice and False for ignore; they define
```

DougBurke replied:
Dang it - I missed this,

DougBurke replied:
Now fixed.

wmclaugh commented on sherpa/tests/test_data.py:
```diff
@@ -1247,3 +1247,120 @@ def test_data1d_notice_errors_out_on_string_range(lo, hi, emsg, ignore):
+
+    expected is the expected response
+    args is a list of 3-tuples of (flag, loval, hival) where
+    flag is True for notice and False for notice; they define
```
```suggestion
    flag is True for notice and False for ignore; they define
```

wmclaugh commented on sherpa/tests/test_data.py:
```diff
@@ -1247,3 +1247,120 @@ def test_data1d_notice_errors_out_on_string_range(lo, hi, emsg, ignore):
+    # Ignore a few points at the start and end
+    d.notice(11, 18)
+
+    # Just check that the default behavior does't change with the filter
```
```suggestion
    # Just check that the default behavior doesn't change with the filter
```

DougBurke commented:
I've added a commit to address @wmclaugh review comments above.

I rebased because we need to pick up recent PRs to let the tests run (because of OTS changes), but this required some rework because the downstream PRs are affected by the merging of #1205. In doing so I noted that our `.gitignore` file was now out of date because of #1218 so I added a commit to fix that too.

DougBurke commented:
Unfortunately this will take a bit of time to get to run the tests as there's quite a back log I generated.
